### ISSUE TITLE
unable to execute tasks after running fortios_system_global with login_banner enabled

### ISSUE THREAD

**Issue by onurbrc** [fortinet-ansible-dev/ansible-galaxy-fortios-collection#77]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
fortios_system_global

##### ANSIBLE VERSION

```
ansible: 2.10.3
fortios_collection: 1.0.15
```
##### PYTHON VERSION
```
Python: 3.8.5
```

##### FORTIGATE VA VERSION

```
v6.2.4 build1112 (GA)
```

##### SUMMARY

 Try to execute Ansible tasks after enabling  disclaimer text banner. 

Workflow actions: 
```
1 - Configure system replacemsg admin with module **fortios_system_replacemsg_admin** ✔️ 
2 - Enabling  **post_login_banner** and **pre_login_banner** with module **fortios_system_global** ✔️ 
3 - Retry step 2 🚫 
4 - Create a local user with module **fortios_user_local** 🚫 

```
##### EXPECTED RESULTS

Execute any new task through API without being blocked by the acceptance of the disclaimer

```
Login Disclaimer
Hello FORTI.
 
**Accept**  | **Decline**

```

##### ACTUAL RESULTS (Execution trace)

Blocked at login step: 

*\"/usr/local/lib/python3.8/dist-packages/fortiosapi/fortiosapi.py\", line 187, in login\n    self._fortiversion = resp_lic['version']\nTypeError: 'Response' object is not subscriptable\n*

```
Including module_utils file ansible_collections/fortinet/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortimanager/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortios/__init__.py
Using module file $PWD/.ansible/collections/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py
<localhost> PUT $PWD/.ansible/tmp/ansible-local-70208bp6elxe/tmp0xaj4p6o TO $PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/AnsiballZ_fortios_system_global.py
<localhost> EXEC /bin/sh -c 'chmod u+x $PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/ $PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/AnsiballZ_fortios_system_global.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python3 $PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/AnsiballZ_fortios_system_global.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r $PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/ > /dev/null 2>&1 && sleep 0'
fatal: [fwl_auto]: FAILED! => {
    "changed": false,
    "module_stderr": "in formatresponse res.content does not exist, should not occur\nTraceback (most recent call last):\n  File \"$PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/AnsiballZ_fortios_system_global.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"$PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/AnsiballZ_fortios_system_global.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"$PWD/.ansible/tmp/ansible-tmp-1607016501.7063422-7024-201866429115598/AnsiballZ_fortios_system_global.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.fortinet.fortios.plugins.modules.fortios_system_global', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python3.8/runpy.py\", line 207, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.8/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.8/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_fortios_system_global_payload_498pcfie/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 2075, in <module>\n  File \"/tmp/ansible_fortios_system_global_payload_498pcfie/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 2055, in main\n  File \"/tmp/ansible_fortios_system_global_payload_498pcfie/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 1427, in login\n  File \"/usr/local/lib/python3.8/dist-packages/fortiosapi/fortiosapi.py\", line 187, in login\n    self._fortiversion = resp_lic['version']\nTypeError: 'Response' object is not subscriptable\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

PLAY RECAP ********************************************************************************************************************************
fwl_auto               : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

```

##### PLAYBOOK/TASKS

Task 1: 
~~~yaml
     - name: Configure system replacemsg admin
       fortios_system_replacemsg_admin:
         host:  "{{ mgmtIp }}"
         username: "{{ login }}"
         password: "{{ password }}"
         vdom:  "{{ ofwlvdom | default('root') }}"
         https: "false"
         ssl_verify: "False"
         state: "present"
         system_replacemsg_admin:
           buffer: "Hello FORTI"
           format: "text"
           header: "none"
           msg_type: "post_admin-disclaimer-text"
       delegate_to: "localhost"
~~~

Task 2: 
~~~yaml
     - name: Configure system replacemsg admin
       fortios_system_global:
         host:  "{{ mgmtIp }}"
         username: "{{ login }}"
         password: "{{ password }}"
         vdom:  "{{ ofwlvdom | default('root') }}"
         https: "false"
         ssl_verify: "False"
         system_global:
           post_login_banner: "enable"
       delegate_to: "localhost"
~~~

![image](https://user-images.githubusercontent.com/7400129/101067441-d84a1680-3597-11eb-80e9-9fc0f1ee7b7f.png)

**Comment by chillancezen** [comment]
hi @onurbrc Bruno,

as of now Ansible doesn't support login with a banner, it just doesn't resolve banner request. 

could you just disable your task 2 for now?

```
config system global
unset post-login-banner
end
```

we don't have quick fix for thepre/post banner issue. 

thanks,
Link

**Comment by onurbrc** [comment]
@chillancezen , 

At work, I must add security banner alerts to respect some standards. 

When the device is in its factory state, I can put this task in the end, at the last position of my deployment, no problem. 
However, when the device is already configured and in use, I find myself stuck.

**Comment by chillancezen** [comment]
I see then.
We will try to make a special patch build for your case in days.

And the official code will be released in next major version.

Thanks,
Link

Sent from my iPhone

On Dec 7, 2020, at 10:31 PM, Bruno Calçado <notifications@github.com<mailto:notifications@github.com>> wrote:


@chillancezen<https://github.com/chillancezen> ,

At work, I must add security banner alerts to respect some standards.

When the device is in its factory state, I can put this task in the end, at the last position of my deployment, no problem.
However, when the device is already configured and in use, I find myself stuck.

**Comment by mbdraks** [comment]
@onurbrc 

Can you please test using fortios_access_token instead of user/pwd?

**Comment by chillancezen** [comment]
> @onurbrc
> 
> Can you please test using fortios_access_token instead of user/pwd?

wow, Michel, how does it happen? access token bypasses banner validation. thank you so much for pointing out this. 

hi @onurbrc ,I've tried with `fortios_access_token`, no longer blocked, please use latest FOS collection with access token based authentication: https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/playbook.html.
```
$ansible-galaxy collection install -f fortinet.fortios
$cat issue77.yml
- hosts: fortigate03
  connection: httpapi
  collections:
  - fortinet.fortios
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443

  tasks:
     - name: Configure system replacemsg admin
       fortios_system_replacemsg_admin:
         vdom:  "root"
         state: "present"
         access_token: "{{ fortios_access_token }}"
         system_replacemsg_admin:
           buffer: "Hello World"
           format: text
           header: none
           msg_type: post_admin-disclaimer-text


     - name: Configure system replacemsg admin
       fortios_system_global:
        access_token: "{{ fortios_access_token }}"
        vdom: 'root'
        system_global:
            post_login_banner: 'enable'
```

thanks, 
Link

**Comment by chillancezen** [comment]
and remember to put your access token in inventory file:
https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/playbook.html#prepare-host-inventory

**Comment by onurbrc** [comment]
Thanks @mbdraks. Nice to know.

@chillancezen How can I programmatically generate this access_token for FOS 6.2.4, please? 

Could you suggest an API request to retrieve this access_token? 

Thanks.

**Comment by mbdraks** [comment]
> wow, Michel, how does it happen? access token bypasses banner validation. thank you so much for pointing out this.

Details can be found on internal bug 0377315, with the refactoring of the login/logout REST API code.

**Comment by chillancezen** [comment]
> Thanks @mbdraks. Nice to know.
> 
> @chillancezen How can I programmatically generate this access_token for FOS 6.2.4, please?
> 
> Could you suggest an API request to retrieve this access_token?
> 
> I understand the steps, but it means that the usage of the Ansible modules is not transparent enough for the users, cause I must be aware of this kind of "bypass" before using the modules. Moreover, if I use an API request to auto generate this access_token, if you change your API, it means that I will be forced to modify my deployment workflow to adapt this additional action. :-(
> 
> Thanks.

yes, I totally understand, Ansible is designed to automate IT infrastructure.

to create an api user, use module: https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/gen/fortios_system_api_user.html
to generate an access token, use monitor API: `/system/api-user/generate-key`

unfortunately, we don't support monitor API in current Ansible collection, we are planning to support all FOS monitor API in next major release. 

here is the workaround to generate api token in Ansible:
```
$cat generate_apitoken.sh
#$1: hostname
#$2 username
#$3 password
#$4 apiuser
curl -s --insecure -X POST https://$1/logincheck -H 'Cache-Control: no-cache' -H 'Content-Type: application/x-www-form-urlencoded' -d 'username=$2&secretkey=$3&ajax=1' -c /tmp/cookies > /dev/null
curl -s -b /tmp/cookies --insecure -X POST https://$1/api/v2/monitor/system/api-user/generate-key -H 'accept: application/json' -H "X-CSRFTOKEN: `cat /tmp/cookies |grep 'ccsrftoken' |grep -v 'ccsrftoken_' |sed 's/\t/ /g' |cut -d ' ' -f 7|tr -d '"'`" -H 'Content-Type: application/x-www-form-urlencoded' -d "{\"api-user\":\"$4\"}"  |grep access_token |tr -d '",' |cut -d ':' -f 2
```

```
$cat generate_apikey.yml
- hosts: localhost
  vars:
    user: 'admin'
    password: 'ca$hc0w'
    host: '192.168.190.161'
    apiuser: foo

  tasks:
   - name: generate token
     script: ./generate_apitoken.sh "{{ host }}" "{{ user }}" "{{ password }}" "{{ apiuser }}"
     register: apitoken


   - debug:
        msg: 'The API token for user:{{apiuser}} is {{apitoken.stdout_lines[0]}}'
```

```
$ansible-playbook -i hosts  generate_apikey.yml

PLAY [localhost] *****************************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [generate token] ************************************************************************************************************************************************************************************************************************
changed: [localhost]

TASK [debug] *********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "msg": "The API token for user:foo is 73nNQHGbq09NGbxz9Hq3s3k94rrprk"
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

```
thanks,
Link

**Comment by onurbrc** [comment]
Hi, 

I'm trying to use fortios_system_api_user but I get "Socket Path Empty! The persistent connection manager is messed up. Try again in a few moments." in the error message. 

I'm using `fortiOS collection 1.1.8` and httpapi as connection plugin.  Here it is the logs:

```
ansible-playbook 2.10.3
  config file = $PWD/ansible/fortinet/ansible.cfg
  configured module search path = ['$PWD/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.8/dist-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 3.8.5 (default, Jul 28 2020, 12:59:40) [GCC 9.3.0]
Using $PWD/ansible/fortinet/ansible.cfg as config file
setting up inventory plugins
host_list declined parsing $PWD/ansible/fortinet/inventory/fortios.ini as it did not pass its verify_file() method
script declined parsing $PWD/ansible/fortinet/inventory/fortios.ini as it did not pass its verify_file() method
auto declined parsing $PWD/ansible/fortinet/inventory/fortios.ini as it did not pass its verify_file() method
yaml declined parsing $PWD/ansible/fortinet/inventory/fortios.ini as it did not pass its verify_file() method
Parsed $PWD/ansible/fortinet/inventory/fortios.ini inventory source with ini plugin
Loading collection fortinet.fortios from $PWD/.ansible/collections/ansible_collections/fortinet/fortios
Loading callback plugin default of type stdout, v2.0 from /usr/local/lib/python3.8/dist-packages/ansible/plugins/callback/default.py
redirecting (type: callback) ansible.builtin.timer to ansible.posix.timer
Loading collection ansible.posix from $PWD/.ansible/collections/ansible_collections/ansible/posix
Attempting to use 'default' callback.
Skipping callback 'default', as we already have a stdout callback.
Attempting to use 'junit' callback.
Attempting to use 'minimal' callback.
Skipping callback 'minimal', as we already have a stdout callback.
Attempting to use 'oneline' callback.
Skipping callback 'oneline', as we already have a stdout callback.
Attempting to use 'tree' callback.
Attempting to use 'timer' callback.
Loading callback plugin ansible.posix.timer of type aggregate, v2.0 from $PWD/.ansible/collections/ansible_collections/ansible/posix/plugins/callback/timer.py

PLAYBOOK: fortios_system_api_user.yaml *************************************************************************************************************************************
Positional arguments: playbooks/fortios_system_api_user.yaml
verbosity: 5
connection: smart
timeout: 10
become_method: sudo
tags: ('all',)
inventory: ('$PWD/ansible/fortinet/inventory/fortios.ini',)
extra_vars: ('targetGroup=FORTIOS',)
forks: 5
1 plays in playbooks/fortios_system_api_user.yaml

PLAY [API User] ************************************************************************************************************************************************************
META: ran handlers

TASK [Configure API users.] ************************************************************************************************************************************************
task path: $PWD/ansible/fortinet/playbooks/fortios_system_api_user.yaml:17
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: gish
<localhost> EXEC /bin/sh -c 'echo ~gish && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo $PWD/.ansible/tmp `"&& mkdir "` echo $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503 `" && echo ansible-tmp-1611305103.2557557-13144-40599251585503="` echo $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503 `" ) && sleep 0'
Including module_utils file ansible/__init__.py
Including module_utils file ansible/module_utils/__init__.py
Including module_utils file ansible/module_utils/basic.py
Including module_utils file ansible/module_utils/_text.py
Including module_utils file ansible/module_utils/common/_collections_compat.py
Including module_utils file ansible/module_utils/common/__init__.py
Including module_utils file ansible/module_utils/common/_json_compat.py
Including module_utils file ansible/module_utils/common/_utils.py
Including module_utils file ansible/module_utils/common/file.py
Including module_utils file ansible/module_utils/common/parameters.py
Including module_utils file ansible/module_utils/common/collections.py
Including module_utils file ansible/module_utils/common/process.py
Including module_utils file ansible/module_utils/common/sys_info.py
Including module_utils file ansible/module_utils/common/text/converters.py
Including module_utils file ansible/module_utils/common/text/__init__.py
Including module_utils file ansible/module_utils/common/text/formatters.py
Including module_utils file ansible/module_utils/common/validation.py
Including module_utils file ansible/module_utils/common/warnings.py
Including module_utils file ansible/module_utils/compat/selectors.py
Including module_utils file ansible/module_utils/compat/__init__.py
Including module_utils file ansible/module_utils/compat/_selectors2.py
Including module_utils file ansible/module_utils/connection.py
Including module_utils file ansible/module_utils/common/json.py
Including module_utils file ansible/module_utils/distro/__init__.py
Including module_utils file ansible/module_utils/distro/_distro.py
Including module_utils file ansible/module_utils/parsing/convert_bool.py
Including module_utils file ansible/module_utils/parsing/__init__.py
Including module_utils file ansible/module_utils/pycompat24.py
Including module_utils file ansible/module_utils/six/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortimanager/common.py
Including module_utils file ansible_collections/__init__.py
Including module_utils file ansible_collections/fortinet/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortimanager/__init__.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py
Including module_utils file ansible_collections/fortinet/fortios/plugins/module_utils/fortios/__init__.py
Using module file $PWD/.ansible/collections/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_api_user.py
<localhost> PUT $PWD/.ansible/tmp/ansible-local-13140ooxoj8do/tmp9_015f61 TO $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503/AnsiballZ_fortios_system_api_user.py
<localhost> EXEC /bin/sh -c 'chmod u+x $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503/ $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503/AnsiballZ_fortios_system_api_user.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python3 $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503/AnsiballZ_fortios_system_api_user.py && sleep 0'
<localhost> EXEC /bin/sh -c 'rm -f -r $PWD/.ansible/tmp/ansible-tmp-1611305103.2557557-13144-40599251585503/ > /dev/null 2>&1 && sleep 0'
fatal: [fwl_auto]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "access_token": null,
            "state": "present",
            "system_api_user": {
                "accprofile": "super_admin",
                "api_key": "adminapi",
                "comments": "test",
                "cors_allow_origin": null,
                "name": "api_user",
                "peer_auth": "enable",
                "peer_group": null,
                "schedule": null,
                "state": null,
                "trusthost": [
                    {
                        "id": "12",
                        "ipv4_trusthost": "0.0.0.0/0",
                        "ipv6_trusthost": "::/0",
                        "type": "ipv4-trusthost"
                    }
                ],
                "vdom": null
            },
            "vdom": "root"
        }
    },
    "msg": "Socket Path Empty! The persistent connection manager is messed up. Try again in a few moments."
}

PLAY RECAP *****************************************************************************************************************************************************************
fwl_auto               : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

Playbook run took 0 days, 0 hours, 0 minutes, 0 seconds
```

I'm using the connection params as explained [here](https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/playbook.html)

Tested with licensed appliance.

**Comment by chillancezen** [comment]
@onurbrc hi, bruno, could you provide your playbook. usually missing something in playbook causes errors like this.

**Comment by onurbrc** [comment]
Yeap... thanks for your help...

Playbook: 

```
---
- name: API User
  hosts: "{{ targetGroup }}"
  connection: httpapi
  gather_facts: no

  collections:
     - fortinet.fortios

  vars:
   vdom: "root"
   ansible_httpapi_validate_certs: no
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_port: 443

  tasks:
    - name: Configure API users.
      fortios_system_api_user:
        vdom:  "{{ vdom }}"
        state: "present"
        system_api_user:
          accprofile: "super_admin"
          api_key: "adminapi"
          comments: "test"
          name: "api_user"
          peer_auth: "enable"
          trusthost:
           - id:  "12"
             ipv4_trusthost: "0.0.0.0/0"
             ipv6_trusthost: "::/0"
             type: "ipv4-trusthost"
      delegate_to: localhost
...
```

**Comment by mbdraks** [comment]
@onurbrc 

You must change the ipv4_trusthost to a valid done (something other than 0.0.0.0/0) and if you are using peer_auth then peer_group must be set, another detail is that delegate_to should also be removed.

It's always interesting to implement the config first using the CLI before using the API, there are some validations that are more explicit using the CLI.

Here's a working playbook using collections 1.1.7/2.0.0 and FOS 6.4:

```yaml
#!/usr/bin/env ansible-playbook
---
- hosts: Enterprise_Core

  vars:
    ansible_connection: httpapi
    ansible_httpapi_use_ssl: yes
    ansible_httpapi_validate_certs: no
    ansible_user: admin_user
    ansible_password: admin_password
    accprofile: super_admin
    ipv4_trusthost: 10.100.55.0/24
    api_user_name: new_api_admin
    trusthost_id: 1
    vdom: root

  tasks:
    - name: Configure API users
      fortinet.fortios.fortios_system_api_user:
        state: absent
        system_api_user:
          name: "{{ api_user_name }}"
          accprofile: "{{ accprofile }}"
          trusthost:
            - id: "{{ trusthost_id }}"
              ipv4_trusthost: "{{ ipv4_trusthost }}"
          vdom:
            - name: "{{ vdom }}"
```

And the output:

```yaml
changed: [Enterprise_Core] => changed=true 
  invocation:
    module_args:
      access_token: null
      state: present
      system_api_user:
        accprofile: super_admin
        api_key: null
        comments: null
        cors_allow_origin: null
        name: new_api_admin
        peer_auth: null
        peer_group: null
        schedule: null
        state: null
        trusthost:
        - id: 1
          ipv4-trusthost: 10.100.55.0/24
        vdom:
        - name: root
      vdom: root
  meta:
    build: 1778
    http_method: POST
    http_status: 200
    mkey: new_api_admin
    name: api-user
    old_revision: d056975b1759381c532661998dfa4585
    path: system
    revision: 822f497a228a9a466c228e8655f99945
    revision_changed: true
    serial: FGVM02TM20000000
    status: success
    vdom: root
    version: v6.4.3
```

**Comment by onurbrc** [comment]
@mbdraks Obrigado

With your suggested adjustments it worked for me now. Thanks. \o/

Knowing that we didn't get a more explicit error message, no idea of what must be changed. Maybe the code could give more tips? 

My bad about the `delegate_to` ,  wrong habit :-/

**Comment by onurbrc** [comment]
Hi, 

Since last week, I'm able to retrieve the api user token, thanks. 

I added the following tasks in the initial playbook right after the generation of the token: 

~~~yaml
     - name: Configure system replacemsg admin
       fortios_system_replacemsg_admin:
         vdom:  "root"
         state: "present"
         access_token: "{{ fortios_access_token }}"
         system_replacemsg_admin:
           buffer: "Hello World"
           format: text
           header: none
           msg_type: post_admin-disclaimer-text

     - name: Enable post login banner
       fortios_system_global:
        access_token: "{{ fortios_access_token }}"
        vdom: 'root'
        system_global:
            post_login_banner: 'enable'
~~~

For the first post_tasks **Configure system replacemsg admin**, there is no problem running once or twice. But the first run of the second task **Enable post login banner** I got an error: 


~~~bash
The full traceback is:
Traceback (most recent call last):
  File "$PWD/.ansible/tmp/ansible-local-53654c2nfluzq/ansible-tmp-1611737987.7222202-53850-101209571558999/AnsiballZ_fortios_system_global.py", line 102, in <module>
    _ansiballz_main()
  File "$PWD/.ansible/tmp/ansible-local-53654c2nfluzq/ansible-tmp-1611737987.7222202-53850-101209571558999/AnsiballZ_fortios_system_global.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "$PWD/.ansible/tmp/ansible-local-53654c2nfluzq/ansible-tmp-1611737987.7222202-53850-101209571558999/AnsiballZ_fortios_system_global.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible_collections.fortinet.fortios.plugins.modules.fortios_system_global', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib/python3.8/runpy.py", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/usr/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py", line 2030, in <module>
  File "/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py", line 2009, in main
  File "/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py", line 1513, in fortios_system
  File "/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py", line 1499, in system_global
  File "/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py", line 181, in set
  File "/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py", line 216, in formatresponse
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
fatal: [fwl_auto]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"$PWD/.ansible/tmp/ansible-local-53654c2nfluzq/ansible-tmp-1611737987.7222202-53850-101209571558999/AnsiballZ_fortios_system_global.py\", line 102, in <module>\n    _ansiballz_main()\n  File \"$PWD/.ansible/tmp/ansible-local-53654c2nfluzq/ansible-tmp-1611737987.7222202-53850-101209571558999/AnsiballZ_fortios_system_global.py\", line 94, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"$PWD/.ansible/tmp/ansible-local-53654c2nfluzq/ansible-tmp-1611737987.7222202-53850-101209571558999/AnsiballZ_fortios_system_global.py\", line 40, in invoke_module\n    runpy.run_module(mod_name='ansible_collections.fortinet.fortios.plugins.modules.fortios_system_global', init_globals=None, run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python3.8/runpy.py\", line 207, in run_module\n    return _run_module_code(code, init_globals, run_name, mod_spec)\n  File \"/usr/lib/python3.8/runpy.py\", line 97, in _run_module_code\n    _run_code(code, mod_globals, init_globals,\n  File \"/usr/lib/python3.8/runpy.py\", line 87, in _run_code\n    exec(code, run_globals)\n  File \"/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 2030, in <module>\n  File \"/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 2009, in main\n  File \"/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 1513, in fortios_system\n  File \"/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/modules/fortios_system_global.py\", line 1499, in system_global\n  File \"/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py\", line 181, in set\n  File \"/tmp/ansible_fortios_system_global_payload_doyzsaji/ansible_fortios_system_global_payload.zip/ansible_collections/fortinet/fortios/plugins/module_utils/fortios/fortios.py\", line 216, in formatresponse\n  File \"/usr/lib/python3.8/json/__init__.py\", line 357, in loads\n    return _default_decoder.decode(s)\n  File \"/usr/lib/python3.8/json/decoder.py\", line 337, in decode\n    obj, end = self.raw_decode(s, idx=_w(s, 0).end())\n  File \"/usr/lib/python3.8/json/decoder.py\", line 355, in raw_decode\n    raise JSONDecodeError(\"Expecting value\", s, err.value) from None\njson.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}

~~~

Any idea? 

Moreover, is there a recommended frequency to retrieve the api_user token? 
Does the token have an expiration time?
For example, if a have a role or a playbook with several tasks, I start the workflow retrieving the token, is this same token valable until the end of that running? 

To avoid this additional steps when we have admin-disclaimer-text , do you have an idea when we will be able to deal with this usecase in a more native/transparent way? 

Thanks.

**Comment by chillancezen** [comment]
> Moreover, is there a recommended frequency to retrieve the api_user token?
>Does the token have an expiration time?
>For example, if a have a role or a playbook with several tasks, I start the workflow retrieving the token, is this same token valable until the end of that running?

No,  a token is valid within at least weeks or months until you regenerate another one.



I ran your case on my testbeds(6.2 and 6.4), not reproducible. I guess it's privilege issue. 
this is the global system configuration which needs highest privilege.
so please make sure the profile of your rest account has that privilege.


for simplicity, I set all r&w privileges for all the targets.
<img width="1919" alt="WX20210127-193806@2x" src="https://user-images.githubusercontent.com/10312789/105986325-be303e80-60d7-11eb-9c14-e5b4d8e06fae.png">

**Comment by onurbrc** [comment]
@chillancezen I'm creating REST API USER with super_admin profile. I have also created a other one with full access.

**Comment by chillancezen** [comment]
yes, below is my playbook,

```
- hosts: fortigate03
  connection: httpapi
  collections:
  - fortinet.fortios
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443
  tasks:
   - name: Configure Firewall Schedule Recurring
     fortios_system_replacemsg_admin:
        vdom:  "{{ vdom }}"
        state: "present"
        access_token: "{{ fortios_access_token }}"
        system_replacemsg_admin:
            buffer: "Hello World"
            format: text
            header: none
            msg_type: post_admin-disclaimer-text

   - name: Enable post login banner
     fortios_system_global:
        access_token: "{{ fortios_access_token }}"
        vdom: 'root'
        system_global:
            post_login_banner: 'enable'
```

no matter how many times I run the playbook, it just passes.


did you run other playbooks after `post_login_banner` is enabled while you didn't put `access_token: "{{ fortios_access_token }}"` in that playbook?

note if you did not issue requests in access_token way, you are going to meet the issue .  https://github.com/fortinet-ansible-dev/ansible-galaxy-fortios-collection/issues/77#issuecomment-740052318

so just remember to put `access_token` into every task.

**Comment by onurbrc** [comment]
Yes, I'm doing the exact same thing and I'm using the token as authentication parameter. 

> did you run other playbooks after post_login_banner is enabled while you didn't put access_token: "{{ fortios_access_token }}" in that playbook?

No, I didn't. I'm running the same playbook with all mentioned tasks inside. If I already know the `fortios_access_token` I do not re-generate one. In this case, I use it during the execution of the same playbook. 

No idea what is happening. 

Thanks.

**Comment by gitnetofr** [comment]
Hi @chillancezen , 

I have the same issue but in my case the devices are not in factory state. They are already deployed and in production. I'd like to start using Ansible to manage configuration. However, knowing that the pre_* and post_* disclaimer banner are already `enabled`, I have an issue to authenticate cause  I only have ansible_user and ansible_password to inform.

What could I do to deal with the connection problem since the devices do not have `fortios_access_token` yet? 

P.S. I want to avoid generate access_token  device per device manually or to disable disclaimer banner only for this token generation. 

Thank you. 
BR.

**Comment by chillancezen** [comment]
> > Thanks @mbdraks. Nice to know.
> > @chillancezen How can I programmatically generate this access_token for FOS 6.2.4, please?
> > Could you suggest an API request to retrieve this access_token?
> > I understand the steps, but it means that the usage of the Ansible modules is not transparent enough for the users, cause I must be aware of this kind of "bypass" before using the modules. Moreover, if I use an API request to auto generate this access_token, if you change your API, it means that I will be forced to modify my deployment workflow to adapt this additional action. :-(
> > Thanks.
> 
> yes, I totally understand, Ansible is designed to automate IT infrastructure.
> 
> to create an api user, use module: https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/gen/fortios_system_api_user.html
> to generate an access token, use monitor API: `/system/api-user/generate-key`
> 
> unfortunately, we don't support monitor API in current Ansible collection, we are planning to support all FOS monitor API in next major release.
> 
> here is the workaround to generate api token in Ansible:
> 
> ```
> $cat generate_apitoken.sh
> #$1: hostname
> #$2 username
> #$3 password
> #$4 apiuser
> curl -s --insecure -X POST https://$1/logincheck -H 'Cache-Control: no-cache' -H 'Content-Type: application/x-www-form-urlencoded' -d 'username=$2&secretkey=$3&ajax=1' -c /tmp/cookies > /dev/null
> curl -s -b /tmp/cookies --insecure -X POST https://$1/api/v2/monitor/system/api-user/generate-key -H 'accept: application/json' -H "X-CSRFTOKEN: `cat /tmp/cookies |grep 'ccsrftoken' |grep -v 'ccsrftoken_' |sed 's/\t/ /g' |cut -d ' ' -f 7|tr -d '"'`" -H 'Content-Type: application/x-www-form-urlencoded' -d "{\"api-user\":\"$4\"}"  |grep access_token |tr -d '",' |cut -d ':' -f 2
> ```
> 
> ```
> $cat generate_apikey.yml
> - hosts: localhost
>   vars:
>     user: 'admin'
>     password: 'ca$hc0w'
>     host: '192.168.190.161'
>     apiuser: foo
> 
>   tasks:
>    - name: generate token
>      script: ./generate_apitoken.sh "{{ host }}" "{{ user }}" "{{ password }}" "{{ apiuser }}"
>      register: apitoken
> 
> 
>    - debug:
>         msg: 'The API token for user:{{apiuser}} is {{apitoken.stdout_lines[0]}}'
> ```
> 
> ```
> $ansible-playbook -i hosts  generate_apikey.yml
> 
> PLAY [localhost] *****************************************************************************************************************************************************************************************************************************
> 
> TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
> ok: [localhost]
> 
> TASK [generate token] ************************************************************************************************************************************************************************************************************************
> changed: [localhost]
> 
> TASK [debug] *********************************************************************************************************************************************************************************************************************************
> ok: [localhost] => {
>     "msg": "The API token for user:foo is 73nNQHGbq09NGbxz9Hq3s3k94rrprk"
> }
> 
> PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
> localhost                  : ok=3    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
> ```
> 
> thanks,
> Link

with FortiOS 2.0.0(https://github.com/fortinet-ansible-dev/ansible-galaxy-fortios-collection/releases/tag/pre-2.0.0), we are able to generate access token with the monitor module.

```
$cat hosts
[fortigates]
fortigate03 ansible_host=192.168.190.160  ansible_user="admin" ansible_password="<your admin password>"
```

```
$cat generate_access_token.yml
- hosts: fortigate03
  connection: httpapi
  collections:
  - fortinet.fortios
  vars:
   vdom: "root"
   ansible_httpapi_use_ssl: yes
   ansible_httpapi_validate_certs: no
   ansible_httpapi_port: 443
  tasks:

   - name: generate access token for api user foo
     fortios_monitor_config:
        vdom: 'root'
        selector: 'generate-key.system.api-user'
        params:
            api-user: 'foo'
```

access token will be returned in the results.

**Comment by chillancezen** [comment]
> Hi @chillancezen ,
> 
> I have the same issue but in my case the devices are not in factory state. They are already deployed and in production. I'd like to start using Ansible to manage configuration. However, knowing that the pre_* and post_* disclaimer banner are already `enabled`, I have an issue to authenticate cause I only have ansible_user and ansible_password to inform.
> 
> What could I do to deal with the connection problem since the devices do not have `fortios_access_token` yet?
> 
> P.S. I want to avoid generate access_token device per device manually or to disable disclaimer banner only for this token generation.
> 
> Thank you.
> BR.

I will have a look tomorrow. thanks.

**Comment by gitnetofr** [comment]
Hi @chillancezen, 

any news? 
:grin:

Thx

**Comment by chillancezen** [comment]
hi @onurbrc @gitnetofr ,

please try latest collection: `2.0.0` or `2.0.1`
does https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-does-ansible-work-with-login-banner solve your issues?

thanks.

**Comment by gitnetofr** [comment]
> hi @onurbrc @gitnetofr ,
> 
> please try latest collection: `2.0.0` or `2.0.1`
> does https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-does-ansible-work-with-login-banner solve your issues?
> 
> thanks.

Hi, I read in your reply in this post [issue 90](https://github.com/fortinet-ansible-dev/ansible-galaxy-fortios-collection/issues/90#issuecomment-816427428) and after checkout the documentation about the [deadlocks](https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-to-safely-generate-access-token), in my case, it will be hard because we have a big amount of fortigates to disable the banner and after generate an api_token. Thanks anyway. 

About the feedback of the new stable releases, I'll test them.

Thank your very much for this new releases!

**Comment by chillancezen** [comment]
maybe we can bypass ansible and disable banners by using raw cli/shell scripting. I will also have a try and let you know later.thanks

Sent from my iPhone

On Apr 9, 2021, at 12:59 AM, gitnetofr ***@***.***> wrote:

﻿

hi @onurbrc<https://github.com/onurbrc> @gitnetofr<https://github.com/gitnetofr> ,

please try latest collection: 2.0.0 or 2.0.1
does https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-does-ansible-work-with-login-banner solve your issues?

thanks.

Hi, I read in your reply in this post issue 90<https://github.com/fortinet-ansible-dev/ansible-galaxy-fortios-collection/issues/90#issuecomment-816427428> and after checkout the documentation about the deadlocks<https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-to-safely-generate-access-token>, in my case, it will be hard because we have a big amount of fortigates to disable the banner and after generate an api_token. Thanks anyway.

About the feedback of the new stable releases, I'll test them.

Thank your very much for this new releases!

**Comment by chillancezen** [comment]
hi @gitnetofr, @onurbrc , 

```
$cat all_fgts.inventory
192.168.190.171####admin####password
192.168.190.172####admin####password
192.168.190.173####admin####password
```

the  script below read the  hosts in `all_fgts.inventory` and disable banners for them:
```
$cat disable_banners.sh
#! /bin/bash

# please make sure sshpass is installed on your host


disable_banner_per_host()
{
    # $1: the host
    # $2: username
    # $3: password
sshpass -p "$3" ssh -o StrictHostKeyChecking=no $2@$1 <<EOF
a
config system global
set pre-login-banner disable
set post-login-banner disable
end
EOF

}

for i in `cat all_fgts.inventory`
do
    args=`echo $i | sed 's/####/ /g'`
    disable_banner_per_host $args
done
```
this script also enable you to turn banner on by changing `disable` to `enable`.
you can also modify the help script and call it in builtin ansible module `shell` or `script`. 
https://docs.ansible.com/ansible/latest/collections/ansible/builtin/shell_module.html


please also make sure `sshpass` is installed on your host.
```
$sudo apt-get  install sshpass
```

mark it `closed`, please feel free to reopen if we can help further.

thanks,
Link.

**Comment by chillancezen** [comment]
@gitnetofr, 

I added this to our faq page. 
there we use pure Ansible playbook to execute raw cli commands. 
https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-to-work-with-raw-fotios-cli

**Comment by gitnetofr** [comment]
> @gitnetofr,
> 
> I added this to our faq page.
> there we use pure Ansible playbook to execute raw cli commands.
> https://ansible-galaxy-fortios-docs.readthedocs.io/en/latest/faq.html#how-to-work-with-raw-fotios-cli

Thank you @chillancezen
I'll share this solution with my colleagues. 😉
